core/shared/src/main/scala/org/typelevel/idna4s/core/uts46/CodePointMapper.scala

```diff
@@ -0,0 +1,220 @@
+            } else {
+              // This is a mapped code point
+              loop(acc.appendCodePoint(otherwise), errors, index + indexIncrement)
+            }
+        }
+      }
+
+    loop(new StringBuilder(len), Nil, 0)
+  }
+
+  def mapCodePoint(codePoint: CodePoint): CodePointStatus = {
+    import CodePointStatus._
+
+    (unsafeMapIntCodePointSentinel0(codePoint.value): @switch) match {
+      // Literals used rather than named constants because scalac won't
+      // generate a switch with the named constants.
+      case -1 => // VALID
+          Valid.always
+      case -2 => // VALID_NV8
+        Valid.nv8
+      case -3 => // VALID_XV8
+        Valid.xv8
+      case -4 => // IGNORED
+        Ignored.instance
+      case -5 => // MAPPED_MULTI_CODE_POINT
+        Mapped.of(unsafeLookupCodePointMultiMapping0(codePoint.value).map(CodePoint.unsafeFromInt))
+      case -6 => // DEVIATION
+        Deviation.one(CodePoint.unsafeFromInt(unsafeLookupDeviationCodePointMapping0(codePoint.value)))
+      case -7 => // DEVIATION_MULTI
+        Deviation.of(unsafeLookupDeviationCodePointMultiMapping0(codePoint.value).map(CodePoint.unsafeFromInt).toList)
+      case -8 => // DISALLOWED
+        Disallowed.instance
+      case -9 => // DISALLOWED_STD3_VALID
+        Disallowed_STD3_Valid.instance
+      case -10 => // DISALLOWED_STD3_MAPPED
+        Disallowed_STD3_Mapped.one(CodePoint.unsafeFromInt(unsafeLookupDisallowedSTD3CodePointMapping0(codePoint.value)))
+      case -11 => // DISALLOWED_STD3_MAPPED_MULTI
+        Disallowed_STD3_Mapped.of(unsafeLookupDisallowedSTD3CodePointMultiMapping0(codePoint.value).map(CodePoint.unsafeFromInt))
+      case -12 => // DEVIATION_IGNORED
+        Deviation.ignored
+      case otherwise =>
+        // Mapped. There is no sentinel value for mapped, because in this case
+        // a code point maps to a single new code point, we just return the
+        // mapping directly.
+        Mapped.one(CodePoint.unsafeFromInt(otherwise))
+    }
+  }
+
+  def unsafeMapIntCodePoint(codePoint: Int): Either[String, CodePointStatus] =
+    CodePoint.fromInt(codePoint).map(mapCodePoint)
+
+  private val ReplacementCharacter: Int =
+    0xFFFD
+
+  sealed abstract class MappingError extends Serializable {
+    def failureIndex: Int
+
+    def message: String
+
+    def codePoint: CodePoint
+
+    override final def toString: String =
+      s"MappingError(message = $message, failureIndex = $failureIndex, codePoint = $codePoint)"
+  }
+
+  object MappingError {
+    private[this] final case class MappingErrorImpl(override val failureIndex: Int, override val message: String, override val codePoint: CodePoint) extends MappingError
+
+    def apply(failureIndex: Int, message: String, codePoint: CodePoint): MappingError =
+      MappingErrorImpl(failureIndex, message, codePoint)
+  }
+
+  sealed abstract class MappingFailure extends RuntimeException with NoStackTrace {
```

Comment:
`Exception`s are conventionally named something like `MappingFailureException` ... right?

Reply:
Sometimes...I'm not sure it is really a standard. See circe's `DecodingFailure` as a counter example. I don't mind changing it though.